Unnamed Update

T-Rizzle12 opened the pull request:
This update is currently not planed to be released any time soon.

Change log so far:
- Some changes to the bots aiming system, “It can now account for weapon recoil.” :tada:
- Made it so if the bot fails to pathfind it will wait 3 seconds before trying again
- Improved what the bot considers on target
- Added more IsValid checks on timers fixing a few Lua errors

T-Rizzle12 commented:
- Attempted to fix bots skipping ladder nodes
- Changed the bot’s pathfinding visibility checks

T-Rizzle12 commented:
- Added two new variables, NextJump and HoldCrouch
- NextJump is when the bot is allowed to press its jump button again 
- HoldCrouch is how long the bot should hold its crouch button
- Added an attempt limit to the path retracing to prevent the game from going into an infinite loop

T-Rizzle12 commented:
- Improved ladder climbing, “fixed a bug where the bot would skip ladder nodes causing it to fall off of ladders”
- Fixed the bot not uncrouching

T-Rizzle12 commented:
- Changed Use to PressUse, because it conflicted with the source code in the game causing it to always be true
- Changed how dropdowns are detected
- Changed some comments
- Improved the no jump cooldown

T-Rizzle12 commented:
- Added and changed some comments
- Added the Entity MetaTable
- Changed the IsDoor method to use the Entity MetaTable

T-Rizzle12 commented:
Function GetInternalVariable is an interface for accessing internal key values on entities. This will allow me to check if a door is open or closed.

T-Rizzle12 commented:
> I’m planning on releasing this update soon, I just need to fix the pathfinding and we should be good to go.

List of what needs to be fixed:
- [ ] Bot can fail to dropdown into holes
- [ x ] Bot will crouch then jump if the upcoming area tells the bot to crouch

T-Rizzle12 commented:
- TRizzleCreateBot will now show an error when you attempt to create a bot without specifying a name
- Made some changes to the function the compiles the list of nav areas to be turned into waypoints for the bot
- Added an IsValid check
- Fixed a logic error on line 1457

T-Rizzle12 commented:
- Changed how the bot creates waypoints
- Temporarily removed the bots ability to skip nodes due to the bot skipping nodes while jumping over gaps causing it to fail the jump
- Fixed the bot constantly recreating the path while on a ladder
- Fixed the bot failing to create nodes after climbing up a ladder
- Temporary removed ShouldDropDown and ShouldJump because they were causing pathing issues

T-Rizzle12 commented:
- Fixed a bug where the bot would crouch while swimming causing it to swim very slowly

T-Rizzle12 commented:
Found this in the CS:GO source code. I need to change these to represent fall damage values in Gmod. It’s very easy to check if realistic fall damage is on.
//--------------------------------------------------------------------------------------------------------------
/**
 * Return approximately how much damage will will take from the given fall height
 */
float CCSBot::GetApproximateFallDamage( float height ) const
{
	// empirically discovered height values
	const float slope = 0.2178f;
	const float intercept = 26.0f;

	float damage = slope * height - intercept;

	if (damage < 0.0f)
		return 0.0f;

	return damage;
}

strubium commented:
> * When the bot doesn’t have enough ammo to fully reload its shotgun it will now reload to how much ammo it has left instead of loading one shell
YES THANK YOU

T-Rizzle12 reviewed (commented):
- Fixed a crouch logic error
- Changed nil checks in the RestoreAmmo Function
- Fixed the bot not properly creating ladder areas
- Readded dropdown detection, "The bot can get stuck sometimes, I am currently working on a fix."
- Fixed an error in visibility checks
- Made the bot jump off of ladders is has finished climbing

T-Rizzle12 reviewed (commented):
- Changed SendBoxedLine’s distance to 36 HU instead of 108 HU
- Fixed a bug in drop-down detection causing the bot not to detect drop-downs when skipping nodes in the path
- Merged the function ClimbUpLadder with Get_Closest_Point due to them doing the same calculations
- Fixed a bug where if the bot repathed while on a ladder it would keep jumping on and off of the ladder, the bot will now only jump off of a ladder once it either reaches the top or bottom of said ladder
- Changed ladder start and end waypoints to the same algorithm valve used with CS:GO bots

T-Rizzle12 commented:
- Changed PointWithinViewAngle to PointWithinCursor since i'm going to use the name for bot LOS in a future update
- Fixed the door check not checking for doors

strubium commented:
> * Changed PointWithinViewAngle to PointWithinCursor since i'm going to use the name for bot LOS in a future update
> * Fixed the door check not checking for doors
> * Changed how the bot aims for the head to be a directly EyePos() instead of EyePos() - Vector( 0, 0, 10 ), “Creating Vectors is expensive in hooks that are run every frame, so this is not only an accuracy improvement, but a code optimization as well”

lmao "Fixed the door check not checking for doors"

fun times, fun times

T-Rizzle12 commented:
> > * Changed PointWithinViewAngle to PointWithinCursor since i'm going to use the name for bot LOS in a future update
> > * Fixed the door check not checking for doors
> > * Changed how the bot aims for the head to be a directly EyePos() instead of EyePos() - Vector( 0, 0, 10 ), “Creating Vectors is expensive in hooks that are run every frame, so this is not only an accuracy improvement, but a code optimization as well”
> 
> lmao "Fixed the door check not checking for doors"
> 
> fun times, fun times

The bot still used doors, but it pressed its use key on any entity it placed its cursor on. It was funny at first, but not what it was intended for. 🤣 
I will be making a way to command the bot to press buttons, but for now it should only use doors.

strubium commented:
> > > * Changed PointWithinViewAngle to PointWithinCursor since i'm going to use the name for bot LOS in a future update
> > > * Fixed the door check not checking for doors
> > > * Changed how the bot aims for the head to be a directly EyePos() instead of EyePos() - Vector( 0, 0, 10 ), “Creating Vectors is expensive in hooks that are run every frame, so this is not only an accuracy improvement, but a code optimization as well”
> > 
> > 
> > lmao "Fixed the door check not checking for doors"
> > fun times, fun times
> 
> The bot still used doors, but it pressed its use key on any entity it placed its cursor on. It was funny at first, but not what it was intended for. 🤣 I will be making a way to command the bot to press buttons, but for now it should only use doors.

Nice! Playing with this bot is probably the most fun I've had in gmod, I really like it!

T-Rizzle12 commented:
> > > > * Changed PointWithinViewAngle to PointWithinCursor since i'm going to use the name for bot LOS in a future update
> > > > * Fixed the door check not checking for doors
> > > > * Changed how the bot aims for the head to be a directly EyePos() instead of EyePos() - Vector( 0, 0, 10 ), “Creating Vectors is expensive in hooks that are run every frame, so this is not only an accuracy improvement, but a code optimization as well”
> > > 
> > > 
> > > lmao "Fixed the door check not checking for doors"
> > > fun times, fun times
> > 
> > 
> > The bot still used doors, but it pressed its use key on any entity it placed its cursor on. It was funny at first, but not what it was intended for. 🤣 I will be making a way to command the bot to press buttons, but for now it should only use doors.
> 
> Nice! Playing with this bot is probably the most fun I've had in gmod, I really like it!

I enjoyed creating this bot, this is some of the most fun I have had coding. I have to thank [Zenlenafelex [PSF]](https://steamcommunity.com/profiles/76561198976669728). If it hadn't of been for their guide on how to program a bot for gmod. This would of taken longer to get to this point.

T-Rizzle12 reviewed (commented):
- Changed EyePos to GetShootPos, “only for player entities”
- Added a vector validation check for SendBoxedLine
- Changed the drop down check to only effect precise drop downs 
- Fixed the bot failing to properly get onto ladders when attempting to climb down them
- Changed how the bot looks at the path it is following to use the same method the bot uses when aiming at its current enemy

T-Rizzle12 commented:
I can use this as another way to check if an enemy is dead 
m_lifeState  ( 0 means the target entity is still alive )

T-Rizzle12 reviewed (commented):
- Added new variable FireWeaponInterval, “this tells the bot the next time it’s allowed to press its IN_ATTACK, instead of depending on math.random.”
- The bot will not press its fire button while reloading
- Fixed the bot’s run and walk speed being slower than a player
- Gave the bot a basic memory interface, “currently not used”
- Another attempt to fix the bot failing to drop-down

strubium commented:
> * Added new variable FireWeaponInterval, “this tells the bot the next time it’s allowed to press its IN_ATTACK, instead of depending on math.random.”
> * The bot will not press its fire button while reloading
> * Fixed the bot’s run and walk speed being slower than a player
> * Gave the bot a basic memory interface, “currently not used”
> * Another attempt to fix the bot failing to drop-down

What could the memory interface do? Could you script actions with it?

strubium commented:
> * Added new variable FireWeaponInterval, “this tells the bot the next time it’s allowed to press its IN_ATTACK, instead of depending on math.random.”
> * The bot will not press its fire button while reloading
> * Fixed the bot’s run and walk speed being slower than a player
> * Gave the bot a basic memory interface, “currently not used”
> * Another attempt to fix the bot failing to drop-down

What could the memory interface do? Could you script actions with it?

T-Rizzle12 commented:
> > * Added new variable FireWeaponInterval, “this tells the bot the next time it’s allowed to press its IN_ATTACK, instead of depending on math.random.”
> > * The bot will not press its fire button while reloading
> > * Fixed the bot’s run and walk speed being slower than a player
> > * Gave the bot a basic memory interface, “currently not used”
> > * Another attempt to fix the bot failing to drop-down
> 
> What could the memory interface do? Could you script actions with it?

Sorry about not being specific with it. The memory interface is for enemies. When the bot sees an enemy it adds to to a list of known enemy’s and the time the bot last saw said enemy. When I implement bot LOS and Hearing this will be important so the bot attacks enemies it has clearly seen before as the bot will no longer be a allowed to see behind it.

strubium commented:
> > > * Added new variable FireWeaponInterval, “this tells the bot the next time it’s allowed to press its IN_ATTACK, instead of depending on math.random.”
> > > * The bot will not press its fire button while reloading
> > > * Fixed the bot’s run and walk speed being slower than a player
> > > * Gave the bot a basic memory interface, “currently not used”
> > > * Another attempt to fix the bot failing to drop-down
> > 
> > 
> > What could the memory interface do? Could you script actions with it?
> 
> Sorry about not being specific with it. The memory interface is for enemies. When the bot sees an enemy it adds to to a list of known enemy’s and the time the bot last saw said enemy. When I implement bot LOS and Hearing this will be important so the bot attacks enemies it has clearly seen before as the bot will no longer be a allowed to see behind it.

Oh ok. That makes sense, thank you!

I didn't know the bot could see behind it lol

T-Rizzle12 reviewed (commented):
- Attempted to fix the bot spamming its reload key
- Removed a few debug print messages 
- Major change to how the bot creates waypoints
- Improved bot drop-down behavior

T-Rizzle12 commented:
- Added fall damage check from CS:GO
- Changed some pathing costs

T-Rizzle12 commented:
- Added some IsValid checks
- Updated comments in code
- Changed drop-down cost multiplier

T-Rizzle12 commented:
Ok, this update is ready for release. 🎉 Just a quick code verification and some typing up of the change notes and we should be good to go.

strubium commented:
Oh Yeah!
